File: Kattis/toy.py
from math import sqrt
from math import fabs

EPS = 1e-7

segments = []
drawing = [None for i in range(100)]
trs = [[] for i in range(10000)]
best = 0
unUsedSegments = 0

def third_point(a, b, c):
    x = (c ** 2 - b ** 2 + a ** 2) / (2 * a)
    y = sqrt(c ** 2 - x ** 2)
    return x, y

def rotate(x, y, m):
    cos2 = 1 / (m ** 2 + 1)
    cos = sqrt(cos2)
    sin = sqrt(1 - cos2)
    if m < 0:
        sin *= -1
    return x * cos - y * sin, x * sin + y * cos

def explore(first = False):
    global best
    global unUsedSegments
    if unUsedSegments < 2:
        return
    goodSegmentsThisTime = []
    for i in segments:
        if not drawing[i]:
            goodSegmentsThisTime.append(i)
    n = len(goodSegmentsThisTime)
    for i in range(n):
        b = goodSegmentsThisTime[i]
        for j in range(i + 1 if first else 0, n):
            c = goodSegmentsThisTime[j]
            if b != c:
                for a in trs[b * 100 + c]:
                    if drawing[a]:
                        A, B = drawing[a]
                        for orientation in [-1, 1]:
                            C = third_point(a, b, c)
                            C = (C[0], C[1] * orientation)
                            dy, dx = B[1] - A[1], B[0] - A[0]
                            if fabs(dx) < EPS:
                                C = -C[1], C[0]
                            else:
                                m = dy / dx
                                C = rotate(C[0], C[1], m)
                            C = (C[0] + A[0], C[1] + A[1])
                            if C[1] < 0:
                                continue
                            if C[1] > best:
                                best = C[1]
                            con1 = (min(A, C), max(A, C))
                            con2 = (min(C, B), max(C, B))
                            drawing[c] = con1
                            drawing[b] = con2
                            unUsedSegments -= 2
                            explore()
                            unUsedSegments += 2
                            drawing[b] = None
                            drawing[c] = None

def check_triangle(a, b, c):
    return a + b > c and a + c > b and b + c > a

if __name__ == '__main__':
    instance = list(map(int, input().split()))
    unUsedSegments = instance[0]
    segments.extend(instance[1:])
    for a in segments:
        for b in segments:
            for c in segments:
                if a != b and b != c and check_triangle(a, b, c):
                    trs[a * 100 + b].append(c)
    for segment in segments:
        drawing[segment] = (0, 0), (segment, 0)
        unUsedSegments -= 1
        explore(True)
        unUsedSegments += 1
        drawing[segment] = None
    print(best)

File: Kattis/toy.cpp
#include <bits/stdc++.h>
using namespace std;

typedef pair<double, double> point;

point nowhere = {-1, -1};
pair<point, point> none = {nowhere, nowhere};

int n, segments[10], unused, ori[2]{-1, 1};
pair<point, point> drawing[100];
vector<int> trs[10000];
double best = 0.0;

inline bool check_triangle(const int& a, const int& b, const int& c){
  return a + b > c && a + c > b && b + c > a;
}

inline point third_point(const double& a, const double& b, const double& c){
  double x = (c * c - b * b + a * a) / (2 * a);
  double y = sqrt(c * c - x * x);
  return {x, y};
}

inline point rotate(const double& x, const double& y, const double& m){
  double cos2 = 1.0 / (m * m + 1);
  double cos = sqrt(cos2);
  double sin = sqrt(1.0 - cos2);
  if (m < 0){
    sin *= -1;
  }
  return {x * cos - y * sin, x * sin + y * cos};
}

void explore(bool first){
  if (unused < 2){
    return;
  }
  for(int i = 0; i < n; i++){
    int b = segments[i];
    if (drawing[b] == none){
      int sj = first ? i + 1 : 0;
      for(int j = sj; j < n; j++){
        int c = segments[j];
        if (b != c && drawing[c] == none){
          for(int a: trs[b * 100 + c]){
            if (drawing[a] != none){
              point A = drawing[a].first;
              point B = drawing[a].second;
              for(int orientation: ori){
                point C = third_point(a, b, c);
                C.second *= orientation;
                double dy = B.second - A.second;
                double dx = B.first - A.first;
                if (fabs(dx) < 1e-7) {
                  C = {-C.second, C.first};
                } else {
                  double m = dy / dx;
                  C = rotate(C.first, C.second, m);
                }
                C.first += A.first;
                C.second += A.second;
                if (C.second < 0) {
                  continue;
                }
                if (C.second > best) {
                  best = C.second;
                }
                pair<point, point> con1 = {min(A, C), max(A, C)};
                pair<point, point> con2 = {min(C, B), max(C, B)};
                drawing[c] = con1;
                drawing[b] = con2;
                unused -= 2;
                explore(false);
                unused += 2;
                drawing[b] = none;
                drawing[c] = none;
              }
            }
          }
        }
      }
    }
  }
}

int main(){
  cin >> n;
  for(int i = 0; i < n; i++){
    cin >> segments[i];
  }
  unused = n;
  for(int a: segments){
    for(int b: segments){
      for(int c: segments){
        if(a != b && b != c && check_triangle(a, b, c)){
          trs[a * 100 + b].push_back(c);
        }
      }
    }
  }
  for(int i = 0; i < n; i++){
    drawing[segments[i]] = none;
  }
  for(int i = 0; i < n; i++){
    drawing[segments[i]] = {{0, 0}, {segments[i], 0}};
    unused--;
    explore(true);
    unused++;
    drawing[segments[i]] = none;
  }
  cout << best << endl;
}